File: Python/utils/examination.py
import copy
import time
from abc import ABC, abstractmethod
from typing import Any, Type


class Logger:
    logs: list[list[Any]] = []
    @classmethod
    def log(cls, *messages) -> None:
        cls.logs.append(copy.deepcopy(list(messages)))
    @classmethod
    def release(cls)->list[Any]:
        logs = cls.logs
        cls.logs = []
        return logs

class TestcaseInterface(ABC):
    @abstractmethod
    def id(self) -> str:
        pass

class TestcaseResult:
    def __init__(self, passed: bool, note: str):
        self.passed = passed
        self.note = note
    @classmethod
    def passed(cls) -> 'TestcaseResult':
        return cls(True, '')
    @classmethod
    def failed(cls, expected: str, actual: str, context: str|None = None) -> 'TestcaseResult':
        note: str = f'{actual} expected {expected}'
        if not context is None:
            note += f' - {context}'
        return cls(False, note)


class ExaminerInterface(ABC):
    @abstractmethod
    def run(self, testcases: tuple[TestcaseInterface, ...]) -> None:
        pass

class FullReviewExaminer(ExaminerInterface, ABC):

    def run(self, testcases: tuple[TestcaseInterface, ...]) -> None:
        print('Start.\n')

        execution_times: list[tuple[TestcaseInterface, float]] = []
        for testcase in testcases:
            start: float = time.perf_counter()
            result: TestcaseResult = self.examine(testcase)
            end: float = time.perf_counter()
            execution_times.append((testcase, end - start))
            if not result.passed:
                print(f'Wrong: {testcase.id()} -> {result.note}')
                for log in Logger.release():
                    print(*log)
                print('')
            Logger.release()

        print('End.')

        print('\nExecution times:')
        total_time: float = 0
        max_time: float = 0
        min_time: float = float('inf')
        for execution_time in execution_times:
            print(f'{execution_time[1]*1_000_000:.2f} μs {execution_time[0].id()}')
            total_time += execution_time[1]
            max_time = max(max_time, execution_time[1])
            min_time = min(min_time, execution_time[1])
        print(f'\nAverage: {total_time/len(execution_times)*1_000_000:.2f} μs')
        print(f'Max: {max_time*1_000_000:.2f} μs')
        print(f'Min: {min_time*1_000_000:.2f} μs')

    @abstractmethod
    def examine(self, testcase: TestcaseInterface) -> TestcaseResult:
        pass

class Validation:
    TYPE_INPUT: str = 'INPUT'
    TYPE_OUTPUT: str = 'OUTPUT'
    VALID_TYPES: tuple[str] = (TYPE_INPUT, TYPE_OUTPUT)
    def __init__(self, type_: str, input_index: int|None=None):
        if type_ not in self.VALID_TYPES:
            raise ValueError(f'Invalid type: {type_}')
        self.type = type_
        self.input_index = input_index
    @classmethod
    def on_input_number(cls, input_index: int) -> 'Validation':
        return cls(cls.TYPE_INPUT, input_index)
    @classmethod
    def on_output(cls) -> 'Validation':
        return cls(cls.TYPE_OUTPUT)
    def should_be_on_input(self) -> bool:
        return self.TYPE_INPUT == self.type

class Testcase(TestcaseInterface):
    def __init__(self, inputs: tuple[Any, ...], expected: Any, validation: Validation=Validation.on_output()) -> None:
        self.inputs = inputs
        self.expected = expected
        self.validation = validation

    def id(self) -> str:
        return str(self.inputs)

class Examiner(FullReviewExaminer):
    def __init__(self, solution: Any, method_name: str):
        self.solution = solution
        self.method_name = method_name

    def examine(self, testcase: Testcase) -> TestcaseResult:
        output: Any = getattr(self.solution, self.method_name)(*testcase.inputs)
        if testcase.validation.should_be_on_input():
            output = testcase.inputs[testcase.validation.input_index]
        if testcase.expected != output:
            return TestcaseResult.failed(testcase.expected, output)
        return TestcaseResult.passed()

class DesignTestcase(TestcaseInterface):
    def __init__(self, calls: list[str], inputs: list[Any], outputs: list[Any]):
        self.calls = calls
        self.inputs = inputs
        self.outputs = outputs
    def id(self) -> str:
        return f'{self.calls} - {self.inputs}'


class DesignExaminer(FullReviewExaminer):
    def __init__(self, solution_class: Type[Any]):
        self.__solution_class = solution_class

    def examine(self, testcase: DesignTestcase) -> TestcaseResult:
        solution: Type[Any] | None = None
        for i, call in enumerate(testcase.calls):
            if self.__solution_class.__name__ == call:
                solution = self.__solution_class(*testcase.inputs[i])
                continue
            output: Any = getattr(solution, call)(*testcase.inputs[i])
            if not testcase.outputs[i] == output:
                return TestcaseResult.failed(testcase.outputs[i], output, f'index {i} {call}({testcase.inputs[i]})')
        return TestcaseResult.passed()

File: TypeScript/utils/examination.ts
import { ListNode } from './linked-list.ts';
import { TreeNode } from './tree.ts';

function deepEquals(a: any, b: any): boolean {
    if (a == null && b == null) return true;
    if (a instanceof ListNode && b instanceof ListNode) return a.equals(b);
    if (a instanceof TreeNode && b instanceof TreeNode) return a.equals(b);
    return JSON.stringify(a) === JSON.stringify(b);
}

export class Logger {
    private static logs: any[][] = [];

    static log(...messages: any[]): void {
        Logger.logs.push(structuredClone(messages));
    }

    static release(): any[][] {
        const logs = Logger.logs;
        Logger.logs = [];
        return logs;
    }
}

export abstract class TestcaseInterface {
    abstract id(): string;
}

export class TestcaseResult {
    readonly passed: boolean;
    readonly note: string;

    private constructor(passed: boolean, note: string) {
        this.passed = passed;
        this.note = note;
    }

    static passed(): TestcaseResult {
        return new TestcaseResult(true, '');
    }

    static failed(expected: any, actual: any, context?: string): TestcaseResult {
        let note = `${String(actual)} expected ${String(expected)}`;
        if (context !== undefined) note += ` - ${context}`;
        return new TestcaseResult(false, note);
    }
}

export abstract class ExaminerInterface {
    abstract run(testcases: readonly TestcaseInterface[]): void;
}

export abstract class FullReviewExaminer extends ExaminerInterface {
    run(testcases: readonly TestcaseInterface[]): void {
        const executionTimes: Array<[TestcaseInterface, number]> = [];
        for (const testcase of testcases) {
            const start = performance.now();
            const result = this.examine(testcase);
            const end = performance.now();
            executionTimes.push([testcase, end - start]);
            if (!result.passed) {
                console.log(`Wrong: ${testcase.id()} -> ${result.note}`);
                for (const log of Logger.release()) {
                    console.log(...log);
                }
                console.log('');
            }
            Logger.release();
        }

        console.log('\nExecution times:');
        let totalTime = 0;
        let maxTime = 0;
        let minTime = Infinity;
        for (const [testcase, elapsed] of executionTimes) {
            console.log(`${(elapsed * 1000).toFixed(2)} μs ${testcase.id()}`);
            totalTime += elapsed;
            if (elapsed > maxTime) maxTime = elapsed;
            if (elapsed < minTime) minTime = elapsed;
        }
        const sorted = [...executionTimes].sort((a, b) => a[1] - b[1]).map(([, t]) => t);
        const n = sorted.length;
        const percentile = (p: number) => sorted[Math.min(Math.floor(p / 100 * n), n - 1)]! * 1000;
        console.log('\n');
        console.log(`p50: ${percentile(50).toFixed(2)} μs`);
        console.log(`p95: ${percentile(95).toFixed(2)} μs`);
        console.log(`Average: ${(totalTime / n * 1000).toFixed(2)} μs`);
    }

    abstract examine(testcase: TestcaseInterface): TestcaseResult;
}

export class Validation {
    static readonly TYPE_INPUT = 'INPUT' as const;
    static readonly TYPE_OUTPUT = 'OUTPUT' as const;

    readonly type: 'INPUT' | 'OUTPUT';
    readonly inputIndex: number | null;

    private constructor(type: 'INPUT' | 'OUTPUT', inputIndex: number | null = null) {
        this.type = type;
        this.inputIndex = inputIndex;
    }

    static onInputNumber(inputIndex: number): Validation {
        return new Validation(Validation.TYPE_INPUT, inputIndex);
    }

    static onOutput(): Validation {
        return new Validation(Validation.TYPE_OUTPUT);
    }

    shouldBeOnInput(): boolean {
        return this.type === Validation.TYPE_INPUT;
    }
}

export class Testcase extends TestcaseInterface {
    readonly inputs: readonly any[];
    readonly expected: any;
    readonly validation: Validation;

    constructor(inputs: readonly any[], expected: any, validation: Validation = Validation.onOutput()) {
        super();
        this.inputs = inputs;
        this.expected = expected;
        this.validation = validation;
    }

    override id(): string {
        return String(this.inputs);
    }
}

export class Examiner extends FullReviewExaminer {
    private readonly fn: (...args: any[]) => any;

    constructor(fn: (...args: any[]) => any) {
        super();
        this.fn = fn;
    }

    override examine(testcase: Testcase): TestcaseResult {
        let output: any = this.fn(...testcase.inputs);
        if (testcase.validation.shouldBeOnInput()) {
            output = testcase.inputs[testcase.validation.inputIndex!];
        }
        if (!deepEquals(testcase.expected, output)) {
            return TestcaseResult.failed(testcase.expected, output);
        }
        return TestcaseResult.passed();
    }
}

export class DesignTestcase extends TestcaseInterface {
    readonly calls: string[];
    readonly inputs: any[][];
    readonly outputs: any[];

    constructor(calls: string[], inputs: any[][], outputs: any[]) {
        super();
        this.calls = calls;
        this.inputs = inputs;
        this.outputs = outputs;
    }

    override id(): string {
        return `${JSON.stringify(this.calls)} - ${JSON.stringify(this.inputs)}`;
    }
}

export class DesignExaminer extends FullReviewExaminer {
    private readonly solutionClass: new (...args: any[]) => any;

    constructor(solutionClass: new (...args: any[]) => any) {
        super();
        this.solutionClass = solutionClass;
    }

    override examine(testcase: DesignTestcase): TestcaseResult {
        let solution: any = null;
        for (let i = 0; i < testcase.calls.length; i++) {
            const call = testcase.calls[i]!;
            const callInputs = testcase.inputs[i] ?? [];
            if (this.solutionClass.name === call) {
                solution = new this.solutionClass(...callInputs);
                continue;
            }
            const output: any = solution[call](...callInputs);
            if (!deepEquals(testcase.outputs[i], output)) {
                return TestcaseResult.failed(testcase.outputs[i], output, `index ${i} ${call}(${JSON.stringify(callInputs)})`);
            }
        }
        return TestcaseResult.passed();
    }
}
